src/Features/CSharp/Portable/UseCompoundAssignment/CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.UseCompoundAssignment;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCompoundAssignment
+{
+    /// <summary>
+    /// Looks for expressions of the form <c>expr ?? (expr = value)</c> and converts it to
+    /// <c>expr ??= value</c>.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUseCompoundCoalesceAssignmentDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseCoalesceCompoundAssignmentDiagnosticId,
+                   CodeStyleOptions.PreferCompoundAssignment,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_compound_assignment), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeCoalesceExpression, SyntaxKind.CoalesceExpression);
+
+        private void AnalyzeCoalesceExpression(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = context.SemanticModel;
+            var coalesceExpression = (BinaryExpressionSyntax)context.Node;
+
+            var option = context.GetOption(CodeStyleOptions.PreferCompoundAssignment, coalesceExpression.Language);
+
+            // Bail immediately if the user has disabled this feature.
+            if (!option.Value)
+                return;
+
+            var coalesceLeft = coalesceExpression.Left;
+            var coalesceReft = coalesceExpression.Right;
+
+            if (!(coalesceReft is ParenthesizedExpressionSyntax parenthesizedExpr))
+                return;
+
+            if (!(parenthesizedExpr.Expression is AssignmentExpressionSyntax assignment))
+                return;
+
+            if (assignment.Kind() != SyntaxKind.SimpleAssignmentExpression)
+                return;
+
+            // have    x ?? (y = z)
```

Comment:
Can `y` here be a ref assignment? If so, do we need tests for it?

Reply:
`y` could indeed be a ref assignment.